MapboxVisionTests/ObservableVideoSourceTests.swift

```diff
@@ -0,0 +1,130 @@
+                                            attributes: [])
+
+    override func setUp() {
+        super.setUp()
+        self.observableVideoSource = ObservableVideoSource()
+    }
+
+    // MARK: - Tests
+
+    func testAddMethodDoesNotThrowWhenIsCalledSerially() {
+        // Given state from setUp()
+
+        // When
+        for _ in 1...Constants.numberOfConcurrentMethodCalls {
+            let observer = VideoSourceObserverMock()
+
+            // Then
+            XCTAssertNoThrow(observableVideoSource.add(observer: observer))
+        }
+    }
+
+    func testRemoveMethodDoesNotThrowWhenIsCalledSerially() {
+        // Given state from setUp()
+
+        // When
+        for _ in 1...Constants.numberOfConcurrentMethodCalls {
+            let observer = VideoSourceObserverMock()
+
+            // Then
+            XCTAssertNoThrow(observableVideoSource.add(observer: observer))
+        }
+    }
+
+    func testNotifyMethodDoesNotThrowWhenIsCalledSerially() {
+        // Given state from setUp()
+
+        // When
+        for _ in 1...Constants.numberOfConcurrentMethodCalls {
+            // Then
+            XCTAssertNoThrow(observableVideoSource.notify { _ in })
+        }
+    }
+
+    func testAddAndNotifyMethodsDoNotThrowWhenAreCalledInParallel() {
+        // Given
+        testExpectation = XCTestExpectation(description: "add(:) and notify() methods handle concurrent method calls properly.")
+        callsCounter = 1
+
+        // When
+        for idx in 1...Constants.numberOfConcurrentMethodCalls {
+            if idx.isMultiple(of: 2) {
+                concurrentQueue.async {
+                    let observer = VideoSourceObserverMock()
+                    XCTAssertNoThrow(self.observableVideoSource.add(observer: observer))
+                    self.incrementCallsCounter()
+                }
+            } else {
+                concurrentQueue.async {
+                    XCTAssertNoThrow(self.observableVideoSource.notify { _ in })
+                    self.incrementCallsCounter()
+                }
+            }
+        }
+
+        // Then
+        wait(for: [testExpectation], timeout: Constants.expectedTestTimeout)
+    }
+
+    func testRemoveAndNotifyMethodsWontCrashWhenAreCalledInParallel() {
+        // Given
+        testExpectation = XCTestExpectation(description: "remove(:) and notify() methods handle concurrent method calls properly.")
+        callsCounter = 1
+
+        for _ in 1...Constants.numberOfConcurrentMethodCalls {
+            let observer = VideoSourceObserverMock()
+            observableVideoSource.add(observer: observer)
+        }
+
+        // When
+        for idx in 1...Constants.numberOfConcurrentMethodCalls {
+            if idx.isMultiple(of: 2) {
+                concurrentQueue.async {
+                    let observer = VideoSourceObserverMock()
+                    XCTAssertNoThrow(self.observableVideoSource.remove(observer: observer))
```

Comment:
if you don't use the generated earlier observers, why do we need to generate them at all

Reply:
Updated.